### ISSUE TITLE
Firefox zoom implementation breaks position: fixed component elements in Docs

### ISSUE THREAD

**Issue by idesigncode** [storybookjs/storybook#16774]
**Describe the bug**
In Firefox, the zoom feature uses `transform` styles (see https://github.com/storybookjs/storybook/pull/12845) - unfortunately this breaks positioning on elements that require `position: fixed` when in the "Docs" view.

Normally a `position: fixed` element's position would correspond to the viewport but in Firefox it corresponds to the element with the `transform` style.

The issue is present regardless of actually using the zoom feature. I have tried to disable the zoom feature (related feature request: https://github.com/storybookjs/storybook/issues/16771) in the hopes that it might also remove the styling used for zoom, however, the styling remains.

**To Reproduce**
I created a simplified reproduction of the issue here: https://github.com/idesigncode/bug-storybook-firefox-position-fixed

To view the issue, open the story in Docs view in both Firefox and another browser.

**System**
  System:
    OS: macOS 12.0.1
    CPU: (8) x64 Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
  Binaries:
    Node: 14.17.3 - ~/.nvm/versions/node/v14.17.3/bin/node
    Yarn: 3.1.0 - /usr/local/bin/yarn
    npm: 7.20.0 - ~/.nvm/versions/node/v14.17.3/bin/npm
  Browsers:
    Chrome: 96.0.4664.55
    Edge: 95.0.1020.53
    Firefox: 94.0.1
    Safari: 15.1
  npmPackages:
    @storybook/addon-actions: ^6.4.0-rc.7 => 6.4.0-rc.7 
    @storybook/addon-docs: ^6.4.0-rc.7 => 6.4.0-rc.7 
    @storybook/addon-essentials: ^6.4.0-rc.7 => 6.4.0-rc.7 
    @storybook/addon-links: ^6.4.0-rc.7 => 6.4.0-rc.7 
    @storybook/react: ^6.4.0-rc.7 => 6.4.0-rc.7 

**Additional context**
My actual use case is documenting a tooltip component that uses fixed positioning to stay within the window:

**Chrome** ✅
![Chrome](https://github.com/idesigncode/bug-storybook-firefox-position-fixed/blob/main/img/Chrome.png?raw=true)

**Firefox** ❌
![Firefox](https://github.com/idesigncode/bug-storybook-firefox-position-fixed/blob/main/img/Firefox.png?raw=true)

**Firefox (with the transform style disabled)** ✅
![Firefox (with the transform style disabled)](https://github.com/idesigncode/bug-storybook-firefox-position-fixed/blob/main/img/Firefox-transform_style_disabled.png?raw=true)

**Comment by Tomastomaslol** [comment]
@idesigncode Thanks for a great bug report! I have done some work on the zoom feature in the past and I might be able to help. 

Going through your examples I can't really see an obvious solution to the problem but I'm going to fire up my development environment over the weekend and see what I can figure out.

@idesigncode Are you interested in figuring out a solution for this issue too?

**Comment by idesigncode** [comment]
Hey @Tomastomaslol, all good, yeah the solution isn't obvious to me either...

For my specific use case I would be happy to disable the zoom feature on the affected stories as long as that also removed the `transform` style (hence the feature request #16771).

Happy to help test and be involved in the discussion but I'm not familiar with the source code.

**Comment by 23thinker** [comment]
Is the story rendered graphically in code?

**Comment by Tomastomaslol** [comment]
I had a look at this and the behaviour is inconsistent across the 2 browsers.

@idesigncode, Could you please explain your use case? To me, the behaviour in firefox looks more correct than in chrome. I would expect the element you are previewing to be positioned relative to the "story wrapper". 

<img width="960" alt="Screen Shot 2021-12-07 at 11 04 59 am" src="https://user-images.githubusercontent.com/1872246/144942205-1d6aae0a-66da-45ac-b036-0c7b88d82afc.png">

In the attached screenshot I added a green box around what I consider to be the "story wrapper." 🙂

**Comment by idesigncode** [comment]
@Tomastomaslol, sure thing:

My use case is documenting a tooltip component that is always visible within the window (if opened). This is achieved by using `position: fixed` and calculating the appropriate top/left values in relation to the viewport.

The calculated values are based off [getBoundingClientRect](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect) which essentially is expecting the first `document` ancestor to be what I would call the "positioning origin".

In Firefox, the first ancestor with the `transform` style kind of becomes the "positioning origin" and so the calculated vertical and horizontal positioning values are no longer correct.

It may be worth noting that all other browsers I've tested this in behave like Chrome so Firefox does appear to be the odd-one-out in regards to this issue.

**Comment by Monk102** [comment]
How can I contribute to this project? I have moderate knowledge of HTML CSS and Javascript!

**Comment by shilman** [comment]
@Monk102 https://storybook.js.org/docs/react/contribute/how-to-contribute

**Comment by prashantneupane09** [comment]
can you guys provide me a simple website code ?😣

**Comment by Radhikakumari001** [comment]
Hey @idesigncode could you please suggest me , how do I start contributing to this project .I am new to the world of open source contribution.

**Comment by idesigncode** [comment]
> Hey @idesigncode could you please suggest me , how do I start contributing to this project .I am new to the world of open source contribution. Please help me

> https://storybook.js.org/docs/react/contribute/how-to-contribute

**Comment by Stroudnat** [comment]
Is this still an active problem?

**Comment by idesigncode** [comment]
@Stroudnat Yes, I'm not aware of any recent changes to resolve this.

**Comment by Stroudnat** [comment]
Thanks @idesigncode just got around to cloning everything and looking at issue repos so Ill check back if I see anything!

**Comment by idesigncode** [comment]
@konsalex sounds interesting... is that in Storybook or the component itself? Can you provide any code examples?

**Comment by konsalex** [comment]
@idesigncode I implemented this in the component it's self. Cannot share code example at the moment, but in a higher level

Before it was a 
```
<ForwardRef>
   <PopoverComponent />
</ForwardRef>
```
which was breaking with [getBoundingClientRect](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect), and the approach of wrapping it with a portal like this,

```tsx
const Portal = ({ children, container }: PortalProps) => {
  return ReactDOM.createPortal(children, container);
};
```

solve the issue

**Comment by nikhilvipparla** [comment]
Test

**Comment by idesigncode** [comment]
I do not expect there is an ideal fix in the Storybook code as the inconsistency lies in the browsers' implementation of the `transform` style...

But if I may propose an alternative: only apply the `transform` style if the zoom feature is not disabled.

I believe the zoom feature is the only reason to use the `transform` style in the first place, so applying it while this feature is disabled seems unnecessary.

**Comment by bodograumann** [comment]
This bug also affects the measuring tool when zooming. That is probably a bigger problem, because for small components zooming when measuring is quite a common workflow.

In chrome it is correctly placed, though blurry:
![Screenshot_20220831-101158](https://user-images.githubusercontent.com/1223583/187629721-931e5108-7d00-4641-81c3-a55955df961e.png)

In firefox it is not scaled correctly and offset:
![Screenshot_20220831-101140](https://user-images.githubusercontent.com/1223583/187629981-f04e7dfe-2be5-4d67-b4af-a0e18b90f685.png)

Recreated in a default react storybook: https://github.com/bodograumann/storybook-zoom-measure

**Comment by dhananjaysa92** [comment]
I would like to look into it and fix. Pls confirm if anybody looking into.

**Comment by dhananjaysa92** [comment]
I started to look into

**Comment by ndelangen** [comment]
Is this still an issue with 7.0 beta?

**Comment by idesigncode** [comment]
@ndelangen Yes, I've just tested by upgrading the reproduction repo above and it is still an issue.

**Comment by stefanpearson** [comment]
👋  I've just encountered this issue and cannot seem to find any workaround! It happens in Chrome as well. It's fine in story mode because the story uses the whole document as a canvas, but has broken in Docs. Strangely, this was working until the recent upgrade of 7.1

**Comment by wfigueroa-chwy** [comment]
This is an issue for us in Chrome as well. Components using `position: fixed` render incorrectly inside the Story doc blocks because of the story wrapper div having a `transform` property in it. This CSS is new on 7.1.

```css
.css-1rj4uy6 {
    height: 68px;
    transform-origin: top left;
    -webkit-transform: scale(1);
    -moz-transform: scale(1);
    -ms-transform: scale(1);
    transform: scale(1);
}
```

<img width="1025" alt="image" src="https://github.com/storybookjs/storybook/assets/64283444/7bc402e4-c6ae-4807-8c4b-fd23cc85e9e9">

**Comment by konsalex** [comment]
@stefanpearson @wfigueroa-chwy indeed that's an issue with `7.1.0`, created a separate issue for this, with reproduction here:
https://github.com/storybookjs/storybook/issues/23586

**Comment by valentinpalkovic** [comment]
This bugfix has introduced the current behaviour: https://github.com/storybookjs/storybook/pull/21303

I'm honestly a little conflicted. On the one hand, I understand your problem, but on the other hand, I believe that the container inside the docs page should be decapsulated from the environment, and elements should not break out of it. Now, you could adjust the height of the container on the Docs page so that modals find a place there as soon as they are triggered. What do you think? Do I overlook something?

**Comment by stefanpearson** [comment]
> 



> This bugfix has introduced the current behaviour: #21303
> 
> I'm honestly a little conflicted. On the one hand, I understand your problem, but on the other hand, I believe that the container inside the docs page should be decapsulated from the environment, and elements should not break out of it. Now, you could adjust the height of the container on the Docs page so that modals find a place there as soon as they are triggered. What do you think? Do I overlook something?

The issue is, or at least in the scenario I've encountered, a popover component that uses position:fixed (e.g. with Floating UI) now loses it's positioning, because transform and fixed positioning do not work together. This means the component is fully broken in docs mode, and not as a fault of the component.

**Comment by konsalex** [comment]
@stefanpearson that's true, but `position: fixed` was inconsistent between browsers also, so some popovers would work in Chrome, but would not work in [Safari](https://github.com/storybookjs/storybook/pull/21303/files#diff-5a5bc0c11b73e81324880a09194634066494c9346e63647f0e902e8d0a9ab803L3)/Firefox .

**Comment by stefanpearson** [comment]
That's problematic because the majority of popovers components in most web apps are implemented with position fixed, either with portalling or with the native dialog element appearing in the top layer. Perhaps this could get better with the arrival of the new Popover API (absolute everything in the top layer?), but alas we're not there yet.

However, the scale transform does make sense for the zoom feature, so I don't expect a rollback. Perhaps there is now a need to disable zoom for individual stories that will certainly break? The only other option we have is to implement the CSS hack outlined by [wfigueroa-chwy](https://github.com/wfigueroa-chwy) above, to ensure our users don't assume our components are broken.

**Comment by Luk-z** [comment]
I'm totally agree with @valentinpalkovic, the correct behavior of the docs container is to simulate a separate viewport.
Consider this scenario: a Modal stories with a lot of fixed/absolute modals in the docs, all modals will be rendered on top of each other.

If you have a story with fixed element you have to consider the height and the position of that fixed element to be correct displayed in the docs.

**Comment by Sidnioulz** [comment]
Having read this thread, I've come to the conclusion that things work as intended right now.

We want fixed popovers to stay within the `Preview` component for each story. If they were positioned according to the entire docs page bounding box, then every story's popover would overlap, and the docs page itself would be unusable. Enforcing a `scale` to reposition `fixed` elements seems correct to me in that regard.

When it comes to JS-based positioning (typically, portals), our only solution is to recommend that users include a portal decorator in their stories. By setting the portal container to be inside the story, you'll ensure portal components sit within the `Preview`.

If there are issues with positioning of portal elements _when zoom is not 100% and when a portal decorator is in use_, please open a separate bug report with a reproduction repository, and we'll look at alternative ways to perform zoom (but we'll likely retain a `scale: 1` in the `Preview` for fixed element anchoring).

Thanks everyone who contributed to this thread. If you think I have made a mistake or disagree with my conclusion, feel free to keep discussing!
